introduce a `@_documentation(...)` attribute to influence SymbolGraphGen

QuietMisdreavus opened the pull request:
Resolves rdar://79049241

This PR introduces a new attribute, `@_documentation(...)`, that influences how SymbolGraphGen handles the symbol it's attached to. There are two forms of this attribute:

- `@_documentation(visibility: ...)` treats the symbol as the given access level in the symbol graph, allowing you to hide a technically-`public` implementation detail from public documentation, or reveal a symbol with an underscored name that would otherwise be hidden.
- `@_documentation(metadata: ...)` adds an arbitrary identifier to the symbol's `"metadata"` field, which can then be handled by other tooling.

This PR also fixes a pre-existing issue where symbols that were declared `internal` in a re-exported module would be included in the symbol graph if it was being requested with an `internal` access level, and similarly for anything below `public`.

QuietMisdreavus committed:
add @_documentation(...) attribute to influence SymbolGraphGen

rdar://79049241

QuietMisdreavus commented:
@swift-ci Please test

franklinsch commented on docs/ReferenceGuides/UnderscoredAttributes.md:
```diff
@@ -482,6 +482,16 @@ Fun fact: Rust has a very similar concept called
 including one called `Send`,
 which inspired the design of `Sendable`.
 
+## `@_nodoc`
```
Should we camel-case this for consistency with the other attributes?

franklinsch commented on include/swift/AST/Module.h:
```diff
@@ -951,7 +951,8 @@ inline SourceLoc extractNearestSourceLoc(const ModuleDecl *mod) {
                                   SmallPtrSetImpl<ModuleDecl *> &Imports,
-                                  llvm::SmallDenseMap<ModuleDecl *, SmallPtrSet<Decl *, 4>, 4> &QualifiedImports);
+                                  llvm::SmallDenseMap<ModuleDecl *, SmallPtrSet<Decl *, 4>, 4> &QualifiedImports,
+                                  bool DropNoDoc = false);
```
I'm fine with this, but for extra flexibility have you also considered making this function take a predicate closure that takes an import and returns whether it should be dropped or not? That way we don't need to hardcode behavior for `@_nodoc` in the implementation.

QuietMisdreavus replied:
This function was effectively written for SymbolGraphGen in the first place, and only used transitively by anything that calls `getDisplayDecls` or `getTopLevelDeclsForDisplay` (which itself calls `getDisplayDecls`). We could change how this is structured, but SymbolGraphGen is still going to be the only one affected. 😅

QuietMisdreavus commented:
Based on discussion in [this Swift Forums thread](https://forums.swift.org/t/nodoc-attribute-for-hiding-symbols-from-the-symbol-graph/59227), i've pushed up a couple commits that change how the attribute works. Instead of a single `@_nodoc` attribute, there is now a general-purpose `@_documentation(...)` attribute that adds an arbitrary "documentation category" to the symbol. This is then read for two special-case categories: `underscored` treats the symbol as having an underscored name, hiding it from symbol graphs and docs by default; and `ignoreUnderscored` does the opposite, taking an underscored symbol and lifting it to its declared visibility (instead of `internal`), allowing public underscored symbols to appear in public docs.

QuietMisdreavus commented:
@swift-ci Please test

QuietMisdreavus commented:
Based on further discussion on the forums, the `@_documentation` attribute now has the following structure:

- `@_documentation(visibility: <level>)` forces the symbol to act like it has the given access level in symbol graphs and documentation, allowing you to override an underscored name or treat a symbol like an internal one if it's only public for implementation-detail reasons.
- `@_documentation(metadata: ...)` adds the given identifier to the symbol graph as a new `"metadata"` field. This can be used to add arbitrary information to a symbol that can be processed by later tooling.

QuietMisdreavus commented:
@swift-ci Please test

franklinsch commented on docs/ReferenceGuides/UnderscoredAttributes.md:
````diff
@@ -117,6 +117,27 @@ extension Text {
 }
 ```
 
+## `@_documentation(metadata: ...)`
````
I assume the value of the `metadata` parameter is just a string that be specified with or without quotes?

QuietMisdreavus replied:
Right now it can only be an identifier, but i could see about making it a quoted string as well.

icanzilb replied:
If it's a string in quotes, developers could stick in useful metadata like JSON, CSV, or hex encoded images which would be fun 👌🏼

franklinsch commented on include/swift/AST/Attr.h:
```diff
@@ -2226,6 +2226,28 @@ class BackDeployAttr: public DeclAttribute {
   }
 };
 
+/// The `@_documentation(...)` attribute, used to note a "category" for a
```
Is the term "category" still correct given the latest changes?

QuietMisdreavus replied:
It's not, good catch 😅

franklinsch commented on docs/ReferenceGuides/UnderscoredAttributes.md:
```diff
@@ -117,6 +117,27 @@ extension Text {
+Adds "documentation metadata" to the symbol. The identifier in the attribute is
+added to the symbol graph in the `"metadata"` field of the symbol. This can be
+used to add an arbitrary grouping or other indicator to symbols for use in
+documentation.
```
Are you able to specify multiple `@_documentation(metadata: …)` attributes?

QuietMisdreavus replied:
I'm not sure; i'll have to check. Offhand, i think the answer is "no" because of checks for duplicate attributes, but i'm not confident about it.

QuietMisdreavus committed:
review: reword comment for DocumentationAttr

QuietMisdreavus committed:
take a quoted string in @_documentation(metadata:)

QuietMisdreavus commented:
I've rebased onto the latest `main`, and squashed down the first few commits to discard the history of the initial implementation (which made rebasing much easier). I also added a tweak to allow quoted strings as doc metadata.

@swift-ci Please test

QuietMisdreavus commented:
@swift-ci Please test

QuietMisdreavus commented:
Rebased to resolve merge conflicts.

@swift-ci Please test

QuietMisdreavus commented:
@swift-ci Please test
